File: javascript/designpatterns/creational/singleton.js
//Singleton pattern 
//Only single object of class is shared in the entire code.

/**
@author : Parixit Singh
*/

//Simple implementation of Singleton pattern in java script
var Singleton = (
    ()=>{
        var instance;
        function createInstance(){
            return {
                first : 1,
                second : 2
            };
        }

        return {
            getInstance: function(){
                if(!instance){
                    instance = createInstance();   
                }
                return instance;
            }
        }
    }
)();


//Singleton implementation using es6 classes

class Singleton1 {
    constructor(){
        if(!Singleton1.instance){
            Singleton1.instance = this;
        }
        return Singleton1.instance;
    }
}

File: java/designpatterns/creational/Singleton.java
//Singleton pattern using Bill Pugh Singleton Method
//Only single object of class is shared in the entire code.

/**
@author : Parixit Singh
*/

package creational;

import java.util.*;


public class Singleton {
    private Singleton(){
        //private constructor
    }

    //Inner class provides singleton instance
    private static class InnerClass {
        private static final Singleton INSTANCE = new Singleton();
    }

    public static Singleton GetInstance(){
        return InnerClass.INSTANCE;
    }
}

/*
//Simple private static member keeping the object and returning it through static method
public class Singleton {
    private static Singleton Instance = new Singleton();
    private Singleton(){
        //private constructor
    }
    public static Singleton GetInstance(){
        return Instance;
    }
}


//On demand initializing a private member with Singleton object and returning it through static method
public class Singleton {
    private static Singleton Instance;
    private Singleton(){
        //private constructor
    }
    public static Singleton GetInstance(){
        if(Instance == null){
            Instance = new Singleton();
        }
        return Instance;
    }
}


//Thread safe Singleton implementation
public class Singleton {
    private static Singleton Instance;
    private Singleton(){
        //private constructor
    }
    public static Singleton GetInstance(){
        if(Instance == null){
            synchronized(Singleton.class){
                if(Instance == null){
                    Instance = new Singleton();
                }
            }
        }
        return Instance;
    }
}

*/